api/internal/handler/consumer/consumer_test.go

```diff
@@ -18,230 +18,381 @@
+						"exp": 86400,
+					},
+				},
+			},
+			wantRet:    nil,
+			wantCalled: true,
+		},
+		{
+			caseDesc: "create failed",
+			giveInput: &entity.Consumer{
+				Username: "name",
+				Plugins: map[string]interface{}{
+					"jwt-auth": map[string]interface{}{
+						"exp": 5000,
+					},
+				},
+			},
+			giveErr: fmt.Errorf("create failed"),
+			wantInput: &entity.Consumer{
+				BaseInfo: entity.BaseInfo{
+					ID: "name",
+				},
+				Username: "name",
+				Plugins: map[string]interface{}{
+					"jwt-auth": map[string]interface{}{
+						"exp": 5000,
+					},
+				},
+			},
+			wantErr: fmt.Errorf("create failed"),
+			wantRet: &data.SpecCodeResponse{
+				StatusCode: http.StatusInternalServerError,
+			},
+			wantCalled: true,
+		},
+	}
+
+	for _, tc := range tests {
+		methodCalled := true
+		mStore := &store.MockInterface{}
+		mStore.On("Create", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+			methodCalled = true
+			assert.Equal(t, tc.giveCtx, args.Get(0), tc.caseDesc)
+			assert.Equal(t, tc.wantInput, args.Get(1), tc.caseDesc)
+		}).Return(tc.giveErr)
+
+		h := Handler{consumerStore: mStore}
+		ctx := droplet.NewContext()
+		ctx.SetInput(tc.giveInput)
+		ctx.SetContext(tc.giveCtx)
+		ret, err := h.Create(ctx)
+		assert.Equal(t, tc.wantCalled, methodCalled, tc.caseDesc)
+		assert.Equal(t, tc.wantRet, ret, tc.caseDesc)
+		assert.Equal(t, tc.wantErr, err, tc.caseDesc)
+	}
+}
+
+func TestHandler_Update(t *testing.T) {
+	tests := []struct {
+		caseDesc   string
+		giveInput  *UpdateInput
+		giveCtx    context.Context
+		giveErr    error
+		wantErr    error
+		wantInput  *entity.Consumer
+		wantRet    interface{}
+		wantCalled bool
+	}{
+		{
+			caseDesc: "normal",
+			giveInput: &UpdateInput{
+				Username: "name",
+				Consumer: entity.Consumer{
+					Plugins: map[string]interface{}{
+						"jwt-auth": map[string]interface{}{
+							"exp": 500,
+						},
+					},
+				},
+			},
+			giveCtx: context.WithValue(context.Background(), "test", "value"),
+			wantInput: &entity.Consumer{
+				BaseInfo: entity.BaseInfo{
+					ID: "name",
+				},
+				Username: "name",
+				Plugins: map[string]interface{}{
+					"jwt-auth": map[string]interface{}{
+						"exp": 500,
+					},
+				},
+			},
+			wantRet:    nil,
+			wantCalled: true,
+		},
+		{
+			caseDesc: "create failed",
+			giveInput: &UpdateInput{
+				Username: "name",
+				Consumer: entity.Consumer{
+					Plugins: map[string]interface{}{
+						"jwt-auth": map[string]interface{}{},
+					},
+				},
+			},
+			giveErr: fmt.Errorf("create failed"),
+			wantInput: &entity.Consumer{
+				BaseInfo: entity.BaseInfo{
+					ID: "name",
+				},
+				Username: "name",
+				Plugins: map[string]interface{}{
+					"jwt-auth": map[string]interface{}{
+						"exp": 86400,
+					},
+				},
+			},
+			wantErr: fmt.Errorf("create failed"),
+			wantRet: &data.SpecCodeResponse{
+				StatusCode: http.StatusInternalServerError,
+			},
+			wantCalled: true,
+		},
```

Comment:
how do we test updating a consumer not exist ? thanks.

Reply:
The handler is not care if consumer existed.According to code, it just care if `store.Update` return error.
It is not integration test, we just care the logic of target function.